File: nuve/nuveClient_ruby/nuve.rb
require 'net/http'
require 'base64'
require 'hmac-sha1'

class Nuve

	def initialize (service, key, url)
		@service = service
		@key = key
		@url = url
	end

	def getRooms ()
		return send('GET', nil, 'rooms', '', '')
	end

    def createRoom (name, options={})
    	return send('POST', {'name' => name, 'options' => options}, 'rooms', '', '')
    end

    def getRoom (room)
 		return send('GET', nil, 'rooms/' + room, '', '')
    end

    def deleteRoom (room)
		return send('DELETE', nil, 'rooms/' + room, '', '')
    end

    def createToken (room, username, role)
    	return send('POST', nil, 'rooms/' + room + '/tokens', username, role)
    end

    def createService (name, key)
    	return send('POST', {'name' => name, 'key' => key}, 'services/', '', '');
    end

    def getServices ()
		return send('GET', nil, 'services/', '', '');
    end

    def getService (service)
   		return send('GET', nil, 'services/' + service, '', '');
    end

    def deleteService (service)
    	return send('DELETE', nil, 'services/' + service, '', '');
    end

    def getUsers (room)
    	return send('GET', nil, 'rooms/' + room + '/users/', '', '');
    end

	private
	def send (method, body, url, username, role)

		uri = URI(@url + url)

		header = 'MAuth realm=http://marte3.dit.upm.es,mauth_signature_method=HMAC_SHA1'

		timestamp = (Time.now.to_f * 1000).to_i
		cnounce = rand(99999)

		toSign = timestamp.to_s + "," + cnounce.to_s

		if username != '' && role != ''

			header += ',mauth_username='
            header +=  username
            header += ',mauth_role='
            header +=  role

            toSign += ',' + username + ',' + role

		end

		signed = calculateSignature(toSign, @key)

		header += ',mauth_serviceid='
        header +=  @service
        header += ',mauth_cnonce='
        header += cnounce.to_s
        header += ',mauth_timestamp='
        header +=  timestamp.to_s
        header += ',mauth_signature='
        header +=  signed

        if body != nil
        	request = Net::HTTP::Post.new(uri.request_uri,  {'Authorization' => header, 'Content-Type' => 'application/json'})
        	request.set_form_data(body)

        else
        	if method == 'GET'
        		request = Net::HTTP::Get.new(uri.request_uri,  {'Authorization' => header})
        	elsif method == 'POST'
        		request = Net::HTTP::Post.new(uri.request_uri,  {'Authorization' => header})
        	else
        		request = Net::HTTP::Delete.new(uri.request_uri,  {'Authorization' => header})
        	end
        end

		Net::HTTP.start(uri.host, uri.port) do |http|
		  response = http.request(request)
		  return response.body
		end

	end

	private
	def calculateSignature (toSign, key)
		hmac = HMAC::SHA1.new(key)
		hex = hmac.update(toSign)
		signed   = Base64.encode64("#{hex}")
		return signed
	end

end

File: nuve/nuveClient_python/nuve.py
import time
import random
import httplib
import json

class Nuve:

    service = None
    key = None
    url = None

    def __init__(self, service, key, url, port):
        self.service = service
        self.key = key
        self.url = url
        self.port = port

    def createRoom(self, name, options, params=None):
        response = self.send('POST', json.dumps({"name": name, "options": options}), "/rooms/", params)
        return response

    def getRooms(self, params=None):
        response = self.send('GET', None, '/rooms/', params)
        return response

    def getRoom(self, room, params=None):
        response = self.send('GET', None, '/rooms/'+room, params)
        return response

    def deleteRoom(self, room, params=None):
        response = self.send('DELETE', None, '/rooms/'+room, params)
        return response

    def createToken(self, room, username, role, params=None):
        response = self.send('POST', json.dumps({}), '/rooms/'+ room + '/tokens', params, username, role)
        return response

    def createService(self, name, key, params=None):
        response = self.send('POST', json.dumps({"name": name, "key": key}), '/services/', params)
        return response

    def getServices(self, params=None):
        response = self.send('GET',  None, '/services/', params)
        return response

    def getService(self, service, params=None):
        response = self.send('GET',  None, '/services/' + service, params)
        return response

    def deleteService(self, service, params=None):
        response = self.send('DELETE',  None, '/services/' + service, params)
        return response

    def getUsers(self, room, params=None):
        response = self.send('GET',  None, '/rooms/' + room + '/users/', params)
        return response

    def getUser(self, room, user, params=None):
        response = self.send('GET',  None, '/rooms/' + room + '/users/' + user, params)
        return response

    def deleteUser(self, room, user, params=None):
        response = self.send('DELETE',  None, '/rooms/' + room + '/users/' + user, params)
        return response


    def send(self, method, body, url, params=None, username="", role=""):
        if (params == None):
            service = self.service
            key = self.key
        else:
            service = params.service
            key = params.key

        timestamp = int(time.time()*1000)
        cnounce = int(random.random()*99999)
        toSign = str(timestamp) + ',' + str(cnounce)

        header = 'MAuth realm=http://marte3.dit.upm.es,mauth_signature_method=HMAC_SHA1'

        if (username != '' and role != ''):
            header += ',mauth_username='
            header +=  username
            header += ',mauth_role='
            header +=  role

            toSign += ',' + username + ',' + role

        signed = self.calculateSignature(toSign, key)

        header += ',mauth_serviceid='
        header +=  service
        header += ',mauth_cnonce='
        header += str(cnounce)
        header += ',mauth_timestamp='
        header +=  str(timestamp)
        header += ',mauth_signature='
        header +=  signed

        conn = httplib.HTTPConnection(self.url, self.port)
        headers = {"Authorization": header, 'Content-Type': 'application/json'}
        conn.request(method, url, body, headers)
        res = conn.getresponse()
        if res.status == 401:
            print res.status, res.reason
            raise Exception('unauthorized')
        response = res.read()
        try:
            data = json.loads(response)
        except Exception:
            data = response
        conn.close()
        return data

    def calculateSignature(self, toSign, key):
        from hashlib import sha1
        import hmac
        import binascii
        hasher = hmac.new(key, toSign, sha1)
        signed = binascii.b2a_base64(hasher.hexdigest())[:-1]
        return signed
